src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +216,249 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
+     * @param methodName the name of the single abstract method
+     * @param methods the information for implementation methods
+     * @return the bytes of the implementation classes
+     */
+    private static byte[] createTemplate(ClassLoader loader, ClassDesc proxyDesc, ClassDesc ifaceDesc,
+                                         String methodName, List<LocalMethodInfo> methods) {
+        return Classfile.of(ClassHierarchyResolverOption.of(ClassHierarchyResolver.ofClassLoading(loader)))
+                .build(proxyDesc, clb -> {
+            clb.withSuperclass(CD_Object);
+            clb.withFlags(ACC_FINAL | ACC_SYNTHETIC);
+            clb.withInterfaceSymbols(ifaceDesc);
+
+            // individual handle fields
+            clb.withField(ORIGINAL_TYPE_NAME, CD_Class, ACC_PRIVATE | ACC_STATIC | ACC_FINAL);
+            clb.withField(ORIGINAL_TARGET_NAME, CD_MethodHandle, ACC_PRIVATE | ACC_FINAL);
+            for (var mi : methods) {
+                clb.withField(mi.fieldName, CD_MethodHandle, ACC_PRIVATE | ACC_FINAL);
+            }
+
+            // <clinit>
+            clb.withMethodBody(CLASS_INIT_NAME, MTD_void, ACC_STATIC, cob -> {
+                cob.constantInstruction(ifaceDesc);
+                cob.putstatic(proxyDesc, ORIGINAL_TYPE_NAME, CD_Class);
+                cob.return_();
+            });
+
+            // <init>(Lookup, MethodHandle originalHandle, MethodHandle implHandle)
+            clb.withMethodBody(INIT_NAME, MTD_void_Lookup_MethodHandle_MethodHandle, 0, cob -> {
+                cob.aload(0);
+                cob.invokespecial(CD_Object, INIT_NAME, MTD_void);
+
+                // WrapperInstance.ensureOriginalLookup
+                cob.aload(1);
+                cob.invokestatic(proxyDesc, "ensureOriginalLookup", MTD_void_Lookup);
+
+                // keep original target
+                // this.originalTarget = originalHandle;
+                cob.aload(0);
+                cob.aload(2);
+                cob.putfield(proxyDesc, ORIGINAL_TARGET_NAME, CD_MethodHandle);
+
+                // convert individual handles
+                for (var mi : methods) {
+                    // this.handleField = implHandle.asType(xxType);
+                    cob.aload(0);
+                    cob.aload(3);
+                    cob.constantInstruction(mi.desc);
+                    cob.invokevirtual(CD_MethodHandle, "asType", MTD_MethodHandle_MethodType);
+                    cob.putfield(proxyDesc, mi.fieldName, CD_MethodHandle);
                 }
+
+                // complete
+                cob.return_();
             });
-        } else {
-            proxy = Proxy.newProxyInstance(proxyLoader,
-                                           new Class<?>[]{ intfc, WrapperInstance.class },
-                                           ih);
-        }
-        return intfc.cast(proxy);
+
+            clb.withMethodBody(ENSURE_ORIGINAL_LOOKUP, MTD_void_Lookup, ACC_PRIVATE | ACC_STATIC, cob -> {
```

Comment:
```suggestion
+            // void ensureOriginalLookup(Lookup) checks if the given Lookup has ORIGINAL
+            // access to this class, i.e. the lookup class is this class; otherwise,
+            // IllegalAccessException is thrown
            clb.withMethodBody(ENSURE_ORIGINAL_LOOKUP, MTD_void_Lookup, ACC_PRIVATE | ACC_STATIC, cob -> {
```